### ISSUE TITLE
Custom resource create might succeed after CRD deletion, leaving data in etcd

### ISSUE THREAD

**Issue by roycaihw** [kubernetes/kubernetes#99181]
Either `/priority important-longterm` or `/priority backlog`

#### What happened:

The chance for the race condition is low. There is a narrow window where CRs can be created after the CRD gets deleted:

1. CR create request hit API server; API server noted the request [isn't terminating yet](https://github.com/kubernetes/kubernetes/blob/2a05c786008f9d0592c0ea51dfb49124bc622100/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L316), went ahead to prepare the handler
2. CRD delete request hit API server; CRD got [marked as terminating](https://github.com/kubernetes/kubernetes/blob/2a05c786008f9d0592c0ea51dfb49124bc622100/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go#L148-L149)
3. CRD finalizer controller deleted existing CRs, removed [CRD cleanup finalizer](https://github.com/kubernetes/kubernetes/blob/2a05c786008f9d0592c0ea51dfb49124bc622100/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/finalizer/crd_finalizer.go#L125-L174); CRD got deleted from etcd
4. API server [served the CR create request](https://github.com/kubernetes/kubernetes/blob/2a05c786008f9d0592c0ea51dfb49124bc622100/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L380), wrote data to etcd
5. destruction of the CR etcd client happened after `#3`, and might [last a long time](https://github.com/kubernetes/kubernetes/blob/2a05c786008f9d0592c0ea51dfb49124bc622100/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L583-L584) to allow `#4` to happen-- as long as `#4` [registered itself to the in-flight requests](https://github.com/kubernetes/kubernetes/blob/2a05c786008f9d0592c0ea51dfb49124bc622100/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L379)

The window is narrow. It is more likely either
- `#4` happened before `#3`-- the created CR got cleaned up by CRD finalizer
- `#4` happened after `#5`-- the create request would get a 503 Service Unavailable "apiserver is shutting down"

#### What you expected to happen:

No CR data persists in etcd after CRD deletion. 

#### How to reproduce it (as minimally and precisely as possible):

- Add `time.Sleep` to make triggering the race easier: https://github.com/roycaihw/kubernetes/commit/7f4d6cbf7ac015edd5dc799c97ab5bc49f814f1c
- `kubectl apply -f crd.yaml` (example yaml: https://github.com/roycaihw/kubernetes/commit/0d3732cab99e79118eeafca0534e1c28dc8c67d9)
- `kubectl apply -f cr.yaml`; immediately `kubectl delete -f crd.yaml` in parallel
- See the CRD was deleted immediately, the CR got 201 CREATED 10s later
- Moreover, re-create the CRD and see the old CR still exists `kubectl apply -f crd.yaml && kubectl get crontabs`

#### Anything else we need to know?:

The narrow window becomes a concern if we increase the latency between `#1` and `#4`, which is what https://github.com/kubernetes/kubernetes/pull/96403 is doing. A mitigation is to read the latest CRD from cache after the increased latency and right before `#4`, to make sure the window is still small: https://github.com/kubernetes/kubernetes/pull/96403/commits/6ddc1eb1fb3c85a241a87ccf66c142be578ff4e3. The race still exist, but it isn't getting worse. 

#### Environment:
- Kubernetes version (use `kubectl version`): HEAD

/sig api-machinery

**Comment by roycaihw** [comment]
I don't have a quick fix. Filed this issue to raise awareness.

/priority important-longterm
/help

**Comment by andrewv22** [comment]
@roycaihw 
To fully mitigate the race condition the only solution I see is:
- Using the lock on the crdHandler before creating the 'create resource handler', thus the deletion of the CRD cannot happen while the creation of the CR is still ongoing (but this might be dumb since it will create an unnecessary dependency, and deletion won't be happening immediately)

Another solution I see is to clean-up etcd after the fact (which seems like a better option code-cleanliness-wise, but could present some issues for scaling). Having a controller (perhaps the garbagecollector) continously iterating over the CRs in etcd and checking if we still have their CRDs.

Let me know what you think. I would like to assign this issue to myself if you think any of these solutions would be the right choice here.

**Comment by roycaihw** [comment]
@andrewv22 Thanks for willing to help. Some thoughts:

> the deletion of the CRD cannot happen while the creation of the CR is still ongoing

CRD and CR can be created by different users (cluster admin v.s. cluster user). I don't think we want block CRD deletion on CR creation, where a CR client (cluster user) may always hold the lock and starve the CRD client (cluster admin).  

> Having a controller (perhaps the garbagecollector) continously iterating over the CRs in etcd and checking if we still have their CRDs

A controller won't be able to watch CRs after the CRD is removed from discovery. Also since the race happens when a CR creation came through right before CR etcd client destruction, a controller that watches the CR creation and sends a deletion will also have to race against the CR etcd client destruction. 

I'd like to see us using in-memory lock to solve the issue in a single server cluster. An etcd distributed lock (an API object?) solution is more expensive. It would be interesting to see if we can solve the issue in an HA cluster as well, but that can be a stretch goal.

**Comment by andrewv22** [comment]
Got it, thanks for the feedback, I'll give it another thought.

**Comment by andrewv22** [comment]
@roycaihw 
Maybe we could solve this using three mutexes:

* data mutex ('M')
* next-to-access mutex ('N')
* low-priority access mutex ('L')

Then modify the functions responsible for CRDs and CRs this way:

* CR Requests (low-priority): lock L, lock N, lock M, unlock N, { do stuff }, unlock M, unlock L
* CRD Requests (high-priority): lock N, lock M, unlock N, { do stuff }, unlock M

This way the CRD requests will always be next in-line.

> CRD and CR can be created by different users (cluster admin v.s. cluster user). I don't think we want block CRD deletion on CR creation, where a CR client (cluster user) may always hold the lock and starve the CRD client (cluster admin).

If the CR creation function has a timeout + the three-mutexes implemented, this shouldn't happen.

**Comment by roycaihw** [comment]
Having a low-priority access mutex ('L') means at least the requests within one custom resource will block each other. Need to reason about what regression we are introducing here (compared to the existing code, where most part of a request's lifecycle is parallelized). We don't want { do stuff } to be the entire lifecycle of a CR request.

**Comment by kagaya85** [comment]
It is true that the cost of using locks is relatively high, perhaps we may solve this problem by using as few locks as possible to guarantee a portion of the creation-deletion order?

Such as using RWmutex to ensure the CRD delete operation **happens before** CR creation, and every CR create operation will check ths CRD status by using Rlock before write to etcd, if the CRD already deleted, return create failed.

The RWmutex need to be writer prefered to avoid starve the CRD client. I will continue to think about whether it is feasible.

**Comment by mdabydeen** [comment]
/assign

I'd like to work on this race condition. Will add a check to reject CR creates once the CRD has begun terminating, plus a regression test, and open a PR shortly.
